agora/Agora/Utils.hs

```diff
@@ -390,3 +409,44 @@ psingletonValue = phoistAcyclic $
         outerTup = pcon $ PMap $ psingleton #$ ppairDataBuiltin # pdata sym # pdata innerTup
         res = pcon $ PValue outerTup
      in res
+
+-- | Finds the TxOut of an effect from TxInfo and TxOutRef
+findTxOutByTxOutRef :: Term s (PTxOutRef :--> PTxInfo :--> PMaybe PTxOut)
+findTxOutByTxOutRef = phoistAcyclic $
+  plam $ \txOutRef txInfo ->
+    pmatch (pfindTxInByTxOutRef # txOutRef # txInfo) $ \case
+      PJust ((pfield @"resolved" #) -> txOut) -> pcon $ PJust txOut
+      PNothing -> pcon PNothing
+
+-- | Get script hash from an Address.
+scriptHashFromAddress :: Term s (PAddress :--> PMaybe PValidatorHash)
+scriptHashFromAddress = phoistAcyclic $
+  plam $ \addr ->
+    pmatch (pfromData $ pfield @"credential" # addr) $ \case
+      PScriptCredential h -> pcon $ PJust $ pfield @"_0" # h
+      _ -> pcon PNothing
+
+-- | Find all TxOuts sent to an Address
+findOutputsToAddress :: Term s (PTxInfo :--> PAddress :--> PBuiltinList PTxOut)
```

Comment:
`PbuiltinList` can't contain non-builtin types. Should return `PbuiltinList (PAsData PtxOut)` here.

Reply:
Oh right 👍🏻